### DISCUSSION TITLE
Restrict Promise.all to only promise-like arguments

### DISCUSSION THREAD

**Discussion by jasikpark** [mattpocock/ts-reset#176] [Ideas]
https://x.com/alvarlagerlof/status/1720182566135779582?s=46&t=G7pFn2184k4yqrwhVaGMiQ

Curious if it would make sense to restrict Promise.all to only accept PromiseLike arguments. You can always Promise.resolve(5) if you really need to place a non promise in it, and then it might avoid this issue? 

probably more properly an eslint change tbh though.

**Comment by yandodov** [comment]
I was about to start the exact same discussion after relizing that I was accidentally slowing down everything using `await` in `Promise.all`. I even wrote [a little post about it](https://dodov.dev/blog/how-to-do-parallelization-right-with-promise-all). TLDR, you can use this tiny wrapper that provides the correct types:

```ts
function parallelize<T extends unknown[]>(promises: {
	[K in keyof T]: Promise<T[K]>
}) {
	return Promise.all(promises)
}
```

…but I wonder if there's a way to change the actual types of `Promise.all`, instead of having to use a wrapper. @mattpocock what do you think? 👀
